Arch arm rework fault handling

ioannisg opened the pull request:
Rework the fault handing for ARM, to separate error reporting from processor fault status resetting and error reason derivation.

Addresses #8950 .

ioannisg committed:
arch: arm: improve documentation of _Fault(.)

This commit improves the documentation of function _Fault(.),
in order to better describe the actual implementation and the
use of the function arguments.
Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

andrewboie commented on arch/arm/core/fault.c:
```diff
@@ -660,64 +649,38 @@ void _Fault(const NANO_ESF *esf, u32_t exc_return)
+			esf = (NANO_ESF *)__TZ_get_PSP_NS();
 			if ((SCB->ICSR & SCB_ICSR_RETTOBASE_Msk) == 0) {
 				PR_EXC("RETTOBASE does not match EXC_RETURN\n");
 				_SysFatalErrorHandler(_NANO_ERR_HW_EXCEPTION, esf);
```
nitpick, not related to this patch: Is it intended that these _SysFatalErrorHandler invocations don't have a return after them?
AFAICT it will be called again after it returns because there is another call site at the end of _Fault. I see five calls to SysFatalErrorHandler in this function, seems like you want a goto instead or something

Also: nothing in this C file should EVER be calling _SysFatalErrorHandler. It should be calling _NanoFatalErrorHandler instead.

andrewboie commented on arch/arm/core/fault.c:
```diff
@@ -554,35 +503,18 @@ static void _ReservedException(const NANO_ESF *esf, int fault)
+	_FaultThreadShow(esf, fault);
+
 	switch (fault) {
 	case 3:
```
nitpick, unrelated to your patch: Are there defined constants we could use instead of just raw numbers for these switch cases?

ioannisg replied:
>nitpick, unrelated to your patch: Are there defined constants we could use instead of just raw numbers for these switch cases?

No, the ARM CMSIS do no include any defined macros for these.

tyrantlizard77 reviewed (changes requested):
This still doesn't work for me.

I basically need this to work in phases:

1) Identify the specific fault
2) Poke whatever bits are necessary to handle the fault condition so that we can continue
3) Have a place to make a policy decision where we decide whether this fault is fatal or not. If not fatal, apply policy and then return. This is where I will check the exception fixup stuff.
4) We didn't return so this is a fatal fault. Dump out interesting information specific to the fault.
5) Call _NanoFatalErrorHandler to dump out the reason code, the faulting thread and other common info, which will then in turn call _SysFatalErrorHandler.

Right now the fault handling is still mixed up with the fault information dumping, and the comments are not correct. For example, MpuFault prints out a bunch of junk, THEN it does things like poke CFSR, and the documentation for the function still states "Dump MPU fault information"

The dumping and the handling need to be separated. Consider what would happen in Linux if stuff was dumped out every time the MMU had a page fault. This whole C file was written under the assumption that these exceptions are always fatal and that's going to be less and less true over time.

ioannisg commented:
Thanks for the feedback, @andrewboi,let's see where we are:

>This still doesn't work for me.

>I basically need this to work in phases:
>
> 1. Identify the specific fault

This is done: `_Fault()` invokes `_FaultHandle()` which identifies the actual processor fault, and handles it accordingly, invoking the corresponding fault handler.

>    2. Poke whatever bits are necessary to handle the fault condition so that we can continue

This is implemented in the various fault handlers, corresponding to each fault condition. Each handler inspects appropriate registers and returns the error reason, that can be used, later, to assess whether we can continue, or not.

> 3. Have a place to make a policy decision where we decide whether this fault is fatal or not. If not fatal, apply policy and then return. This is where I will check the exception fixup stuff.

For this, I have a comment. This should not be done in the generic fault handler, but, instead, inside the specific ones. This is because, the applied policies shall be _only_ applicable to specific fault errors.

> 4. We didn't return so this is a fatal fault. Dump out interesting information specific to the fault.
    Call _NanoFatalErrorHandler to dump out the reason code, the faulting thread and other common info, which will then in turn call _SysFatalErrorHandler.

This is done, now, at the end of _Fault(). For now, we still call the _sysFatalErrorHandler(), but I have tracked the change to _Nano_ in #8950.

ioannisg committed:
arch: arm: Call NanoFatalErrorHandler and split out Secure stack dump

This commit implements two changes.
- It splits out the dumping of the Secure stack information to a
  separate function. Dumping will only occur if the highest dump
  level has been selected.
- Forces _Fault() to invoke _NanoFatalErrorHandler, instead of
  _SysFatalErrorHandler. In addition, early return statemets are
  implemented in _Fault(), so the fatal error handler is only
  invoked once.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: re-organize arm fault handling

This commit does several things:
- It unifies the way the fault inspection is implemented,
  irrespective of the selected fault dumping level.
- It removes the dumping of the Thread ID and faulting
  instruction address, as these are now displayed by
  _NanoFatalErrorHandler.
- It introduces a new debug print directive, to be used
  for the highest dumping level (2).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: allow processor to ignore/recover from faults

This commit adds the implementation that allows the ARM CPU
to recover from (otherwise fatal) MPU faults. A new error
reason, _NANO_ERR_RECOVERABLE, is introduced. The error
reason is used to suppress fault dump information, if the
error is actually recoverable.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

andrewboie commented:
This is better.
However, for the example of the patch in #8863 I am getting bus faults and not MPU faults if I don't handle the exception:

```
***** BUS FAULT *****
  Executing thread ID (thread): 0x20402590
  Faulting instruction address:  0x4020c2
  Precise data bus error
  Address: 0xfffffff0
Fatal fault in thread 0x20402590! Aborting.
```

So this needs to be for bus and not mpu faults. Actually it might need to be both:

1) Bus faults for truly nonsense memory addresses
2) MPU faults for memory addresses in a MPU region which is defined as not readable in kernel mode, like a guard page or something

Given that the ranges in the exception table are so far just a single instruction, this is why I found it a lot simpler to just check the exception table on *any* fault since we know what the instruction is doing.

tyrantlizard77 reviewed (approved):
I'm going to merge this as is and see if I can tweak it to get the right fault handling conditions I need.

ioannisg commented:
@andrewboie you are right. This work needs to be completed by doing similar things for Bus and HardFaults:
- BusFaults in case of NXP MPU, which generates BusFaults instead of MPU Faults,
- HardFaults for ARMv6-M/ARMv8-M_Baseline which does not have other fault exceptions
These exceptions  need to have a way to by-pass not-fatal errors, similarly to what I did, here for MPU faults.

>Given that the ranges in the exception table are so far just a single instruction, this is why I found it a lot simpler to just check the exception table on any fault since we know what the instruction is doing

I am a bit confused here. Anyways, I think it is better to have individual code blocks for checking for non-fatal (recoverable) errors at each fault exception, in order to have better control of what faults we can assess as recoverable and what we cannot.

For example, for ARMv7-M SOCs with ARM MPU we shall be expecting MPU faults, so we shall not try to recover from BusFaults. Vice-versa for SOCs with NXP MPU. Also, for ARMv7-M, with MPU (ARM, or NXP) enabled, we shall not expect to recover from Hard Faults, as we should expect Bus or MMU faults instead.
